### PR TITLE
Install samples sources

### PR DESCRIPTION
**Pull request by mshabunin**
### This pullrequest changes

- unified samples binary generation (`BUILD_EXAMPLES`) and sources installation (`INSTALL_C_EXAMPLES`)
- fixed several compilation warnings in samples and OpenGL code
- replaced broad "opencv2/opencv.hpp" includes with exact modules
- allowed building installed samples sources as standalone project (`cmake <install>/share/OpenCV/samples && make`)

**Note**: test with related _opencv_contrib_ patch opencv/opencv_contrib#1550

```
force_builders=Custom
docker_image:Custom=ubuntu-cuda:16.04
allow_multiple_commits=1
```

### CODE REVIEW

**Comment by sturkmen72** [comment]
maybe related this PR

dnn examples have the prefix "example_dnn-" like example_dnn-caffe_googlenet.exe

it should be "dnn-example-" to be uniform with other samples.

**Comment by alalek** [comment]
I agree that names should be consistent.
But "category" should be first: `opencv_test_dnn` / `opencv_perf_dnn` / `example_dnn_*`

**Comment by mshabunin** [comment]
@sturkmen72 , this PR unifies all samples names across the library on all platforms. DNN samples will be named `example_dnn_...`

**Comment by catree** [comment]
Related (resolves?) #7614

**Comment by LaurentBerger** [comment]
Does it solve https://github.com/opencv/opencv/issues/9784

**Comment by mshabunin** [comment]
>Related (resolves?) #7614

Yes.

>Does it solve #9784 ?

@LaurentBerger , not sure. I will try to reproduce this issue with the patch.

**Comment by mshabunin** [comment]
@LaurentBerger , I've added the fix for #9784.

File: samples/cpp/CMakeLists.txt [hunk 1]
```diff
@@ -1,117 +1,65 @@
...
+project(cpp_samples)
+ocv_include_modules_recurse(${OPENCV_CPP_SAMPLES_REQUIRED_DEPS})
+file(GLOB_RECURSE cpp_samples RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} *.cpp)
+if(NOT HAVE_OPENGL)
+  ocv_list_filterout(cpp_samples Qt_sample)
+endif()
+if(NOT HAVE_opencv_cudaarithm OR NOT HAVE_opencv_cudafilters)
+  ocv_list_filterout(cpp_samples "/gpu/")
+endif()
+if(NOT VTK_USE_FILE)
```
Comment (alalek) [inline samples/cpp/CMakeLists.txt]: This condition should check "opencv_viz" module build status too

See this [build failure](http://pullrequest.opencv.org/buildbot/builders/master-mac-static/builds/10368)

**Pull request merged**